packages/graphql-types-generator/src/swift/appSyncCompatibilityTypes.ts

```diff
@@ -0,0 +1,550 @@
+      }
+    }
+    return jsonArray
+  }
+}
+
+public typealias GraphQLMap = [String: JSONEncodable?]
+
+extension Dictionary where Key == String, Value == JSONEncodable? {
+  public var withNilValuesRemoved: Dictionary<String, JSONEncodable> {
+    var filtered = Dictionary<String, JSONEncodable>(minimumCapacity: count)
+    for (key, value) in self {
+      if value != nil {
+        filtered[key] = value
+      }
+    }
+    return filtered
+  }
+}
+
+public protocol GraphQLMapConvertible: JSONEncodable {
+  var graphQLMap: GraphQLMap { get }
+}
+
+public extension GraphQLMapConvertible {
+  var jsonValue: Any {
+    return graphQLMap.withNilValuesRemoved.jsonValue
+  }
+}
+
+public extension GraphQLMapConvertible {
+  func evaluate(with variables: [String: JSONEncodable]?) throws -> Any {
+    return try graphQLMap.evaluate(with: variables)
+  }
+}
+
+public typealias GraphQLID = String
+
+public protocol GraphQLOperation: AnyObject {
+  static var rootCacheKey: String { get }
+  
+  static var operationString: String { get }
+  static var requestString: String { get }
+  static var operationIdentifier: String? { get }
+  
+  var variables: GraphQLMap? { get }
+  
+  associatedtype Data: GraphQLSelectionSet
+}
+
+public extension GraphQLOperation {
+  static var requestString: String {
+    return operationString
+  }
+
+  static var operationIdentifier: String? {
+    return nil
+  }
+
+  var variables: GraphQLMap? {
+    return nil
+  }
+}
+
+public protocol GraphQLQuery: GraphQLOperation {}
+public extension GraphQLQuery {
+  static var rootCacheKey: String { return "QUERY_ROOT" }
+}
+
+public protocol GraphQLMutation: GraphQLOperation {}
+public extension GraphQLMutation {
+  static var rootCacheKey: String { return "MUTATION_ROOT" }
+}
+
+public protocol GraphQLSubscription: GraphQLOperation {}
+
+public extension GraphQLSubscription {
+    static var rootCacheKey: String { return "SUBSCRIPTION_ROOT" }
+}
+
+public protocol GraphQLFragment: GraphQLSelectionSet {
+  static var possibleTypes: [String] { get }
+}
+
+public extension GraphQLOperation {
+    static func getResponseGraphQLSelections() -> [GraphQLSelection] {
+        return Data.selections
+    }
+}
+
+public typealias Snapshot = [String: Any?]
+
+public protocol GraphQLSelectionSet: Decodable {
+  static var selections: [GraphQLSelection] { get }
+  
+  var snapshot: Snapshot { get }
+  init(snapshot: Snapshot)
+}
+
+extension GraphQLSelectionSet {
+    public init(from decoder: Decoder) throws {
+        if let jsonObject = try? JSONValue(from: decoder) {
+            let encoder = JSONEncoder()
+            let jsonData = try encoder.encode(jsonObject)
+            let decodedDictionary = try JSONSerialization.jsonObject(with: jsonData, options: []) as! [String: Any]
+            let optionalDictionary = decodedDictionary.mapValues { $0 as Any? }
+
+            self.init(snapshot: optionalDictionary)
+        } else {
+            self.init(snapshot: [:])
+        }
```

Comment:
This is where the custom decoder was added. It uses a self defined JSONValue type to get the data into an JSONValue object before serializing and deserializing back to the expected output a `Snapshot` (typealias for [String:Any?])